From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Paul Praschl <praschlpaul@g-p.at>
Date: Wed, 22 Jul 2026 12:44:48 +0200
Subject: [PATCH 018/117] serdev: add stop-bits configuration

Serdev clients could select baud rate, parity and flow control, but not
the number of stop bits required by some devices.

Add a serdev stop-bit API and wire it through tty_port termios handling.
Existing clients retain one-stop-bit behavior unless they explicitly opt
in.

Signed-off-by: Paul Praschl <praschlpaul@g-p.at>

Orchard-replay-of: aed6d003403b628e4905b9e3c650283f2d60a872
---
 drivers/tty/serdev/core.c           | 12 ++++++++++++
 drivers/tty/serdev/serdev-ttyport.c | 28 ++++++++++++++++++++++++++++
 include/linux/serdev.h              | 10 ++++++++++
 3 files changed, 50 insertions(+)

diff --git a/drivers/tty/serdev/core.c b/drivers/tty/serdev/core.c
index 7500efcdf..2901574bd 100644
--- a/drivers/tty/serdev/core.c
+++ b/drivers/tty/serdev/core.c
@@ -351,6 +351,18 @@ int serdev_device_set_parity(struct serdev_device *serdev,
 }
 EXPORT_SYMBOL_GPL(serdev_device_set_parity);
 
+int serdev_device_set_stopbits(struct serdev_device *serdev,
+			       enum serdev_stopbits stopbits)
+{
+	struct serdev_controller *ctrl = serdev->ctrl;
+
+	if (!ctrl || !ctrl->ops->set_stopbits)
+		return -EOPNOTSUPP;
+
+	return ctrl->ops->set_stopbits(ctrl, stopbits);
+}
+EXPORT_SYMBOL_GPL(serdev_device_set_stopbits);
+
 void serdev_device_wait_until_sent(struct serdev_device *serdev, long timeout)
 {
 	struct serdev_controller *ctrl = serdev->ctrl;
diff --git a/drivers/tty/serdev/serdev-ttyport.c b/drivers/tty/serdev/serdev-ttyport.c
index bab1b143b..5acb898a7 100644
--- a/drivers/tty/serdev/serdev-ttyport.c
+++ b/drivers/tty/serdev/serdev-ttyport.c
@@ -207,6 +207,33 @@ static int ttyport_set_parity(struct serdev_controller *ctrl,
 	return 0;
 }
 
+static int ttyport_set_stopbits(struct serdev_controller *ctrl,
+				enum serdev_stopbits stopbits)
+{
+	struct serport *serport = serdev_controller_get_drvdata(ctrl);
+	struct tty_struct *tty = serport->tty;
+	struct ktermios ktermios = tty->termios;
+
+	switch (stopbits) {
+	case SERDEV_STOPBITS_1:
+		ktermios.c_cflag &= ~CSTOPB;
+		break;
+	case SERDEV_STOPBITS_2:
+		ktermios.c_cflag |= CSTOPB;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	tty_set_termios(tty, &ktermios);
+
+	if ((tty->termios.c_cflag & CSTOPB) !=
+	    (ktermios.c_cflag & CSTOPB))
+		return -EINVAL;
+
+	return 0;
+}
+
 static void ttyport_wait_until_sent(struct serdev_controller *ctrl, long timeout)
 {
 	struct serport *serport = serdev_controller_get_drvdata(ctrl);
@@ -255,6 +282,7 @@ static const struct serdev_controller_ops ctrl_ops = {
 	.close = ttyport_close,
 	.set_flow_control = ttyport_set_flow_control,
 	.set_parity = ttyport_set_parity,
+	.set_stopbits = ttyport_set_stopbits,
 	.set_baudrate = ttyport_set_baudrate,
 	.wait_until_sent = ttyport_wait_until_sent,
 	.get_tiocm = ttyport_get_tiocm,
diff --git a/include/linux/serdev.h b/include/linux/serdev.h
index b6c3d957e..94ab14b66 100644
--- a/include/linux/serdev.h
+++ b/include/linux/serdev.h
@@ -74,6 +74,11 @@ enum serdev_parity {
 	SERDEV_PARITY_ODD,
 };
 
+enum serdev_stopbits {
+	SERDEV_STOPBITS_1,
+	SERDEV_STOPBITS_2,
+};
+
 /*
  * serdev controller structures
  */
@@ -84,6 +89,8 @@ struct serdev_controller_ops {
 	void (*close)(struct serdev_controller *);
 	void (*set_flow_control)(struct serdev_controller *, bool);
 	int (*set_parity)(struct serdev_controller *, enum serdev_parity);
+	int (*set_stopbits)(struct serdev_controller *ctrl,
+			    enum serdev_stopbits stopbits);
 	unsigned int (*set_baudrate)(struct serdev_controller *, unsigned int);
 	void (*wait_until_sent)(struct serdev_controller *, long);
 	int (*get_tiocm)(struct serdev_controller *);
@@ -253,6 +260,7 @@ static inline int serdev_device_set_tiocm(struct serdev_device *serdev, int set,
 {
 	return -EOPNOTSUPP;
 }
+
 static inline int serdev_device_break_ctl(struct serdev_device *serdev, int break_state)
 {
 	return -EOPNOTSUPP;
@@ -294,6 +302,8 @@ static inline int serdev_device_set_rts(struct serdev_device *serdev, bool enabl
 
 int serdev_device_set_parity(struct serdev_device *serdev,
 			     enum serdev_parity parity);
+int serdev_device_set_stopbits(struct serdev_device *serdev,
+			       enum serdev_stopbits stopbits);
 
 /*
  * serdev hooks into TTY core
